pkg/clusteragent/custommetrics/provider.go

```diff
@@ -70,70 +149,27 @@ func (p *datadogProvider) ListAllMetrics() []provider.CustomMetricInfo {
-		for _, in := range p.externalMetrics {
-			externalMetricsInfoList = append(externalMetricsInfoList, in.info)
-		}
-		return externalMetricsInfoList
+	if !p.serving {
+		return nil
 	}
-
-	log.Debugf("Local copy of external metrics from the global store is outdated by %d seconds, resyncing now", copyAge-p.maxAge)
-	rawMetrics, err := p.store.ListAllExternalMetricValues()
-	if err != nil {
-		log.Errorf("Could not list the external metrics in the store: %s", err.Error())
-		return externalMetricsInfoList
-	}
-
-	for _, metric := range rawMetrics {
-		// Only metrics that exist in Datadog and available are eligible to be evaluated in the HPA process.
-		if !metric.Valid {
-			continue
-		}
-		var extMetric externalMetric
-		extMetric.info = provider.ExternalMetricInfo{
-			Metric: metric.MetricName,
-		}
-		// Avoid overflowing when trying to get a 10^3 precision
-		q, err := resource.ParseQuantity(fmt.Sprintf("%v", metric.Value))
-		if err != nil {
-			log.Errorf("Could not parse the metric value: %v into the exponential format", metric.Value)
-			continue
-		}
-		extMetric.value = external_metrics.ExternalMetricValue{
-			MetricName:   metric.MetricName,
-			MetricLabels: metric.Labels,
-			Value:        q,
-		}
-		externalMetricsList = append(externalMetricsList, extMetric)
-
-		externalMetricsInfoList = append(externalMetricsInfoList, provider.ExternalMetricInfo{
-			Metric: metric.MetricName,
-		})
+	var externalMetricsInfoList []provider.ExternalMetricInfo
+	log.Tracef("Listing available metrics as of %s", time.Unix(p.timestamp, 0).Format(time.RFC850))
+	for _, metric := range p.externalMetrics {
+		externalMetricsInfoList = append(externalMetricsInfoList, metric.info)
 	}
-	p.externalMetrics = externalMetricsList
-	p.timestamp = metav1.Now().Unix()
-	log.Debugf("ListAllExternalMetrics returns %d metrics", len(externalMetricsInfoList))
 	return externalMetricsInfoList
 }
 
-// GetExternalMetric is called every 30 seconds as a result of:
-// - The registering of the External Metrics Provider
-// - The creation of a HPA manifest with an External metrics type.
-// - The validation of the metrics against Datadog
-// Every replica answering to a ListAllExternalMetrics will populate its cache with a copy of the global cache.
-// If the copy does not exist or is too old (>1 HPA controller default run cycle) we refresh it.
+// GetExternalMetric is called by the PodAutoscaler Controller to get the value of the external metric it is currently evaluating.
 func (p *datadogProvider) GetExternalMetric(namespace string, metricSelector labels.Selector, info provider.ExternalMetricInfo) (*external_metrics.ExternalMetricValueList, error) {
-	matchingMetrics := []external_metrics.ExternalMetricValue{}
 
-	p.ListAllExternalMetrics() // get up to date values from the cache or the Global Store
+	if time.Now().Unix()-p.timestamp > 2*p.maxAge || !p.serving {
```

Comment:
We can probably do it the other way around so it's cheaper if `!p.serving`

Reply:
true